Add tests for geostore creation based on points

tiagojsag opened the pull request.

thomas-maschler commented:
@tiagojsag This should normally fix the issue. However, E2E tests still fail. The carto module throws an error. The repair geometry agojsag method makes an external request to Carto. Normally this should be mocked. Do the tests block external requests?
I have issues running the service locally, somehow the service doesn't register with control tower in my local setup. So I drill down to the bottom of why the test still doesn't pass.

tiagojsag commented:
Hi @thomas-maschler 

Indeed, there is a carto error. This is due to the carto user not being defined in the configuration, which I now fixed. The tests using docker are already using a fake value for this variable and if you are running the test using `npm`, you need to set the `CARTODB_USER` env var to any value you want - the value of the var does not affect the outcome of the tests.

Answering another question you raised: I was not blocking requests to external services, but now I am. I also added mocks for the carto calls currently being made by our tests, so that those calls stop failing the tests. Keep in mind that I did not validate the response of those mocks, so it's possible you have to change/recreate them. You can do so using this: https://github.com/nock/nock#recording 

As for registering to Control Tower, you probably need to adjust the CT_URL variable value, and this is also dependent on how you are firing up this MS. We can chat more on slack.

Thanks

thomas-maschler commented:
Hey @tiagojsag ,
tests pass locally. It looks like the carto user is not set in TravisCI. I am note sure where to do that.
Please also note, that i wasn't abel to try it out myself. Somehow my Control Tower gives me a headache again. When trying to register geostore I now get 
```
control-tower |   <-- POST /api/v1/microservice
control-tower | [2019-02-21T03:19:26.913Z]  INFO: API-Gateway-DEV/32 on 39f9d82ecf56: Registering microservice
control-tower | [2019-02-21T03:19:26.922Z]  INFO: API-Gateway-DEV/32 on 39f9d82ecf56: Registering new microservice with name GFW GeoStore API and url http://172.21.182.178:3100
control-tower | [2019-02-21T03:19:26.922Z] DEBUG: API-Gateway-DEV/32 on 39f9d82ecf56: Search if microservice already exist
control-tower | [2019-02-21T03:19:26.926Z] DEBUG: API-Gateway-DEV/32 on 39f9d82ecf56: Creating microservice with status pending
control-tower | Thu, 21 Feb 2019 03:19:26 GMT stadistics-plugin Saving stadistic
control-tower | [2019-02-21T03:19:26.942Z] ERROR: API-Gateway-DEV/32 on 39f9d82ecf56: document must have an _id before saving
control-tower |     MongooseError: document must have an _id before saving
control-tower |         at new MongooseError (/opt/control-tower/node_modules/mongoose/lib/error/mongooseError.js:14:11)
control-tower |         at Timeout._onTimeout (/opt/control-tower/node_modules/mongoose/lib/model.js:257:18)
control-tower |         at ontimeout (timers.js:458:11)
control-tower |         at tryOnTimeout (timers.js:296:5)
control-tower |         at Timer.listOnTimeout (timers.js:259:5)
control-tower | [2019-02-21T03:19:26.943Z] ERROR: API-Gateway-DEV/32 on 39f9d82ecf56: Cannot set property 'status' of null
control-tower |     TypeError: Cannot set property 'status' of null
control-tower |         at Function.register (/opt/control-tower/app/src/services/microservice.service.js:313:30)
control-tower |         at <anonymous>
control-tower |         at process._tickCallback (internal/process/next_tick.js:160:7)
control-tower | [2019-02-21T03:19:26.980Z]  INFO: API-Gateway-DEV/32 on 39f9d82ecf56: Time request ==>  76 msec
control-tower |   --> POST /api/v1/microservice 204 76ms 
```

Any hint how to fix that?
